File: leetcode/ugly_number.py
# https://leetcode.com/problems/ugly-number/

#  Write a program to check whether a given number is an ugly number.
# 
# Ugly numbers are positive numbers whose prime factors only include 2, 3, 5. For example, 6, 8 are ugly while 14 is not ugly since it includes another prime factor 7.
# 
# Note that 1 is typically treated as an ugly number. 

# Easiest way to do this: factorize the number. If at any point in time you come across a factor that isn't 2/3/5, ret 

import math
class Solution:
    def isUgly(self, num):
        uglyFactors = {2: True, 3: True, 5: True}
        def factorize(number):
            if number == 1:
                return True
            for i in range(2, int(math.floor(number / 2)+1)):
                remainder = number % i
                if remainder == 0:
                    if i not in uglyFactors:
                        return False
                    return factorize(int(number / i))

            if number not in uglyFactors:
                return False
            return True

        return factorize(num)


sol = Solution()
print(sol.isUgly(214797179))
print(sol.isUgly(14))
print(sol.isUgly(6))
print(sol.isUgly(8))

File: leetcode/es6/uglyNumber.js
// https://leetcode.com/problems/ugly-number/
//
// Write a program to check whether a given number is an ugly number.

// Ugly numbers are positive numbers whose prime factors only include 2, 3, 5. For example, 6, 8 are ugly while 14 is not ugly since it includes another prime factor 7.
// 
// Note that 1 is typically treated as an ugly number. 

/**
 * @param {number} num
 * @return {boolean}
 */

// New approach: remove all factors 2, 3, and 5, then see if there are any other prime factors
var isUgly = function(num) {
  var uglyFactors = {
    2: true,
    3: true,
    5: true
  };

  // If all factors were ugly, it's ugly
  if (num === 1) {
    return true;
  } else if (num === 0) {
    return false;
  }

  for (var factor in uglyFactors) {
    // If there are any ugly factors, remove and recurse
    if (num % factor === 0) {
      return isUgly(num / factor);
    } 
  }
  // Otherwise, we have a non-ugly number
  return false;
};

// This version works (I believe) but is too slow
// var isUgly = function(num) {
//   var uglyFactors = {
//     2: true,
//     3: true,
//     5: true
//   }
// 
//   var hasNonUglyFactor = function(num) {
//     var isPrime = true;
//     for (var i = 2; i <= Math.sqrt(num); i++) {
//       if (num % i === 0) {
//         isPrime = false;
//         if (hasNonUglyFactor(num / i)) {
//           return true;
//         }
//       }
//     }
//     if (isPrime && !uglyFactors[num]) { 
//       return true;
//     }
//     return false;
//   }
// 
//   return ! hasNonUglyFactor(num); 
// }

console.log(isUgly(4));
console.log(isUgly(10));
console.log(isUgly(20));
console.log(isUgly(14));
console.log(isUgly(70));
console.log(isUgly(905391974));
console.log(isUgly(2123366400));




